# A first look at the data

## Introduction

In this section we will set the paths and load the data. We will then explain the data structure and do a few sanity checks. Finally follow some examples for how to perform simples plots of the raw data. 

Import the required modules:

In [1]:
import os
import os.path as op
import openneuro 

import numpy as np
import mne
from mne_bids import BIDSPath, read_raw_bids

## Loading the data

Loading the data. The MEGIN data are stored in the FIF format which is a binary format with embedded labels.

As a first step, we set the path to the data. Note that this will dependent on where you have stored the dataset. Afterwards, we set the file names.

In [2]:
dataset = 'ds004346'

# data_path = r'THE PATH TO WHERE YOU WANT TO DOWNLOAD THE DATA'
bids_root = r'Z:\Projects\FLUX\MEG_data\openneuro'

# download the dataset from openneuro
openneuro.download(dataset=dataset, target_dir=bids_root)


👋 Hello! This is openneuro-py 2022.2.0. Great to see you! 🤗

   👉 Please report problems 🤯 and bugs 🪲 at
      https://github.com/hoechenberger/openneuro-py/issues

🌍 Preparing to download ds004346 …


📁 Traversing directories for ds004346 : 0 entities [00:00, ? entities/s]

📥 Retrieving up to 349 files (5 concurrent downloads). 
✅ Finished downloading ds004346.
 
🧠 Please enjoy your brains.
 


Now we need to prepare the data for reading. For this purpose we need to creat an mne_bids.BIDSPath which is the workhorse object of MNE-BIDS.

In [5]:
subject = '01'
session = '01'
task = 'SpAtt'  # name of the task
run = '01'  # we have two runs for this subject, let's look at run = 1
suffix = 'meg'

bids_path = BIDSPath(subject=subject, session=session,
task=task, run=run, suffix=suffix, root=bids_root)

There is a limit to the file size of the FIF files. If the MEG recordings exceed this limit, the MEGIN acquisition system automatically split the data in two or more FIF files. In those cases, by reading the first FIF file MNE-Python you will automatically read all the linked split files. In our case, the sample dataset was broken into 2 sub-files by the operator and we need to read them one by one. 

We now created the BIDSPath object, let's read the first meg file:

In [6]:
data1 = read_raw_bids(bids_path=bids_path, verbose=False)

C:\Users\ghafarit\AppData\Local\Temp\ipykernel_17560\210554171.py:1: RuntimeWarning: The unit for channel(s) STI001, STI002, STI003, STI004, STI005, STI006, STI007, STI008, STI009, STI010, STI011, STI012, STI013, STI014, STI015, STI016, STI101, STI102 has changed from V to NA.
  data1 = read_raw_bids(bids_path=bids_path, verbose=False)


To get some basic information from the FIF file write:

In [7]:
print(data1)

<Raw | sub-01_ses-01_task-SpAtt_run-01_meg.fif, 343 x 1457000 (1457.0 s), ~737 kB, data not loaded>


The 5.4 MB refer to the size of the data set. The 1457.0 s is the recording time. As the sampling rate was 1000 Hz, this results in 1457000 samples recorded in 343 channels. These channels include the MEG, EOG, triggers channels etc.

To get some additional information:

In [8]:
print(data1.info)

<Info | 21 non-empty values
 acq_pars: ACQactiveGround 1 ACQch.BIO001.gain 2000 ACQch.BIO001.highpass ...
 bads: []
 ch_names: EOG001, EOG002, ECG003, IASX+, IASX-, IASY+, IASY-, IASZ+, ...
 chs: 6 misc, 11 IAS, 102 Magnetometers, 204 Gradiometers, 18 Stimulus, 2 SYST
 custom_ref_applied: False
 description: Anonymized using a time shift to preserve age at acquisition
 dev_head_t: MEG device -> head transform
 dig: 403 items (3 Cardinal, 4 HPI, 396 Extra)
 events: 1 item (list)
 experimenter: mne_anonymize
 file_id: 4 items (dict)
 gantry_angle: 68.0
 highpass: 0.1 Hz
 hpi_results: 1 item (list)
 line_freq: 50.0
 lowpass: 330.0 Hz
 meas_date: 2000-01-01 00:00:00 UTC
 meas_id: 4 items (dict)
 nchan: 343
 proj_name: mne_anonymize
 projs: generated with autossp-1.2.0: off, generated with autossp-1.2.0: ...
 sfreq: 1000.0 Hz
 subject_info: 4 items (dict)
>


This set of information show important parameters of the dataset. For instance, we can see that the gantry was positioned at 68 degress (*gantry_angle*) and that the sampling frequency was 1000 Hz (*sfreq*). The data were lowpass filtered at 330.0 Hz (*lowpass*) and highpass filtered at 0.1 Hz (*highpass*) prior to the digital sampling.

**Question 1:** What is the purpose of the 330 Hz lowpas-filter (hint: see Analog Filters for Data Conversion, Chapter 3; Smith et al., 1999)

**Question 2:** What is the purpose of 0.1 Hz highpass-filter?


The data object (*data1*) allows for a simple inspection of the data by plotting the spectra:

In [9]:
%matplotlib inline
data1.plot_psd(fmax=60).show()

NOTE: plot_psd() is a legacy function. New code should use .compute_psd().plot().
Effective window size : 0.256 (s)


c:\Users\ghafarit\Anaconda3\envs\flux\lib\site-packages\mne\viz\utils.py:137: UserWarning: Matplotlib is currently using module://matplotlib_inline.backend_inline, which is a non-GUI backend, so cannot show the figure.
  (fig or plt).show(**kwargs)


Note the 10 Hz alpha activity on the gradiometers as well as the 50 Hz line noise. The frequency of the line noise is 50 Hz in European countries including Russia whereas it is 60 in South Korea as well as South and North America. In Japan it can be either 50 or 60 Hz depending on the region.

**Question 3:** Why is the line-noise 50 Hz larger for the magnetometers than the gradiometers?


To show the raw data use:

In [13]:
%matplotlib inline
data1.plot(duration=10,title='Raw').show()

Channels marked as bad:
none
Channels marked as bad:
none
Channels marked as bad:
none


To enable the interactive functions of the plotting tool: 

In [11]:
%matplotlib qt
data1.plot(duration=10,title='Raw')

Waiting for Loading-Thread to finish... (max. 10 sec)
Channels marked as bad:
none


This will open the plot in a new window. Use the arrow keys to move over channels and time. Click on the help button for more information.

**Question 3:** Scroll through the MEG data and identify the following artifacts (include figures in the response):

- Eye-blink
- A muscle contraction (characterized by high-frequency activity
- The cardiac artifact



## Preregistration and publication

Preregistration: report the sampling frequency and the properties of the anti-aliasing lowpass filter. Also report the peripheral data to be recorded (e.g. the EOG and ECG)

Publication, example:

"The ongoing MEG data were recorded using the TRIUX system from MEGIN. This system has 102 magnetometers and 204 planer gradiometers. These are placed at 102 locations each having one magnetometer and a set of two orthogonal gradiometers. The horizontal and vertical EOG data as well as the ECG were acquired together with the MEG data. The data were sampled at 1000 Hz and stored for offline analysis. Prior to sampling, a lowpass filter at ~330 Hz was applied. To record the horizontal EOG, a pair of electrodes were attached approximately 2.5 cm away from the outer canthus of each eye. To record the vertical EOG, a pair of electrodes were placed above and below the right eye in line with the pupil. The ECG was recorded from a pair of electrodes placed on the left and right collarbone. Four head position indicator coils (HPIs) were placed behind the left and right ear as well as on the left and right forehead just below the hairline. The positions of the HPIs, the nasion, the left and right preauricular points, as well as the surface points of the scalp, were digitized using a PolhemusTM device. "

## References
Smith, W.S. (1997) The Scientist and Engineer's Guide to Digital Signal Processing. California Technical Publishing. ISBN 0-9660176-3-3 [Online version](http://www.dspguide.com/)